File: android/src/main/java/com/smileidentity/react/viewmanagers/SmileIDBiometricKYCViewManager.kt
package com.smileidentity.react.viewmanagers

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.module.annotations.ReactModule
import com.smileidentity.react.toConsentInfo
import com.smileidentity.react.toIdInfo
import com.smileidentity.react.utils.getBoolOrDefault
import com.smileidentity.react.utils.getImmutableMapOrDefault
import com.smileidentity.react.utils.getMapOrDefault
import com.smileidentity.react.utils.getStringOrDefault
import com.smileidentity.react.views.SmileIDBiometricKYCView

@ReactModule(name = SmileIDBiometricKYCViewManager.NAME)
class SmileIDBiometricKYCViewManager(
  private val reactApplicationContext: ReactApplicationContext
) : BaseSmileIDViewManager<SmileIDBiometricKYCView>(reactApplicationContext) {

  override fun getName(): String = NAME

  override fun createSmileView(): SmileIDBiometricKYCView {
    if (smileIDView == null) {
      smileIDView = SmileIDBiometricKYCView(reactApplicationContext)
    }
    return smileIDView as SmileIDBiometricKYCView
  }

  override fun applyArgs(view: SmileIDBiometricKYCView, args: ReadableMap?) {
    args?.let {
      val idInfoMap = it.getMap("idInfo")
        ?: return view.emitFailure(IllegalArgumentException("idInfo is required to run Biometric KYC"))
      val consentInformationMap = it.getMap("consentInformation")
        ?: return view.emitFailure(IllegalArgumentException("consentInformation is required to run Biometric KYC"))
      val idInfo = idInfoMap.toIdInfo()
      view.consentInformation = consentInformationMap.toConsentInfo()
      view.extraPartnerParams = it.getImmutableMapOrDefault("extraPartnerParams")
      view.userId = it.getStringOrDefault("userId")
      view.jobId = it.getStringOrDefault("jobId")
      view.idInfo = idInfo
      view.allowAgentMode = it.getBoolOrDefault("allowAgentMode", false)
      view.showAttribution = it.getBoolOrDefault("showAttribution", true)
      view.showInstructions = it.getBoolOrDefault("showInstructions", true)
      view.allowNewEnroll = it.getBoolOrDefault("allowNewEnroll", false)
      view.renderContent()
    }
  }

  companion object {
    const val NAME = "SmileIDBiometricKYCView"
    const val COMMAND_SET_PARAMS = 6
  }
}

File: ios/ViewManagers/SmileIDBiometricKYCViewManager.swift
import Foundation
import React
import SmileID
import SwiftUI

@objc(SmileIDBiometricKYCViewManager)
class SmileIDBiometricKYCViewManager: SmileIDBaseViewManager {
  override func getView() -> UIView {
    BaseSmileIDView(frame: .zero, contentView: AnyView(SmileIDBiometricKYCView(product: product, smileIDUIViewDelegate: self)), product: product)
  }
  
  @objc func setParams(_ node: NSNumber, commandId _: NSNumber, params: NSDictionary) {
    /*  UI Updates on the Main Thread:async ensures that the UI update is scheduled to run on the next cycle of the run loop, preventing any potential blocking of the UI if the update were to take a noticeable amount of time
     */
    DispatchQueue.main.async {
      if let component = self.bridge.uiManager.view(forReactTag: node) as? BaseSmileIDView {
        let onResult = params["onResult"] as? RCTDirectEventBlock
        guard let idInfo = params["idInfo"] as? NSDictionary else {
          onResult?(["error": SmileIDError.unknown("idInfo is required to run Biometric KYC")])
          return
        }
        if let consentInformation = params["consentInformation"] as? NSDictionary{
          self.product.consentInformation = consentInformation.toConsentInfo()
        } else  {
          self.product.consentInformation = ConsentInformation(
            consentGrantedDate: getCurrentIsoTimestamp(),
            personalDetailsConsentGranted: false,
            contactInformationConsentGranted: false,
            documentInformationConsentGranted: false
          )
        }
        self.product.extraPartnerParams = params["extraPartnerParams"] as? [String: String] ?? [:]
        self.product.userId = params["userId"] as? String
        self.product.jobId = params["jobId"] as? String
        self.product.allowNewEnroll = params["allowNewEnroll"] as? Bool ?? false
        self.product.allowAgentMode = params["allowAgentMode"] as? Bool ?? false
        self.product.showAttribution = params["showAttribution"] as? Bool ?? true
        self.product.showInstructions = params["showInstructions"] as? Bool ?? true
        self.product.useStrictMode = params["useStrictMode"] as? Bool ?? false
        self.product.idInfo = idInfo.toIdInfo()
        self.product.onResult = onResult
      }
    }
  }
}
